modules/scrollback/index.js

```diff
@@ -110,108 +111,402 @@ module.exports.formatLine = function(line) {
+ *    then the first line matching /cat/
+ *
+ * joe,bob 3
+ * -> Return the third line from joe, then the third line from bob.
+ *
+ * joe, /dog/, 3
+ * -> Return the first line from joe, then the first line matching /dog/,
+ *    then the third line.
+ *
+ * joe 3 bob /dog/ /cat/ bill, 4
+ * -> Return the third line from joe, then the first line from bob
+ *    matching /dog/, then the first line matching /cat/, then the
+ *    first line from bill, then the fourth line.
+ */
+var STATES = Object.freeze({
+  NICK: 1,
+  REGEX: 2,
+  LINES: 3,
+  NEW: 4,
+  ERROR: 5
+});
+
+module.exports.parseSpecs = function(input) {
+  var specs = [];
+  var spec = {nicks: [], regexes: [], lines: []};
+
+  // State machine variables
+  //var state = ''; // one of 'nick', 'regex', 'lines', 'new', 'error'
+  var error = '';
+
+  var i = 0;
+  var result = findNextState(input, i, state);
+  i = result.index;
 
-module.exports.getFormattedScrollbackLinesFromRanges = function(channel, ranges, cb) {
-  var linesToGet = [];
-  var parts = ranges;
-  if(parts.length === 0 || parts.length === 1 && parts[1] === "") {
-    parts = ["1"];
+  if(result.state === STATES.NEW) { // Can't start in 'new'
+    return {error: 'Invalid leading characters'};
   }
+  var state = result.state;
 
-  var i;
+  // Parse input into tokens
+  while(i < input.length) {
+    if(state === STATES.NICK) {
+      result = parseNick(input, i);
 
-  for(i=0;i<parts.length;i++) {
-    // TODO, there's some serious duplication going on here
-    if(/^(-|\d)/.test(parts[i])) {
-      // Starts with a - or number, can't be a nick
-      linesToGet.push(rangeParser.parse(parts[i]));
-    } else if(/^\/.*\/$/.test(parts[i])) {
-      // sandwitched between '/'s, it's a regex
-      try {
-        var regex = new RegExp(parts[i].substring(1, parts[i].length - 1));
-        linesToGet.push(regex);
-      } catch(ex) {
-        return cb("Could not parse regex: " + parts[i] + ", " + ex.toString());
+      if(i != result.index) { // Got a nick
+        i = result.index;
+        spec.nicks.push(result.result);
       }
-    } else {
-      // This is a nick. If the next one is a number then it's for this nick
-      var lineObj = {
-        from: parts[i]
-      };
-      i++;
-      if(/^(-|\d)/.test(parts[i])) {
-        lineObj.lines = rangeParser.parse(parts[i]);
-      } else if(/^\/.*\/$/.test(parts[i])) {
-        // sandwitched between '/'s, it's a regex
-        try {
-          var regex = new RegExp(parts[i].substring(1, parts[i].length - 1));
-          lineObj.regex = regex;
-        } catch(ex) {
-          return cb("Could not parse regex: " + parts[i] + ", " + ex.toString());
-        }
-      } else {
-        lineObj.lines = [1];
-        i--;
-        // Backtrack, nick not followed by anything it turns out
+
+    }
+    else if(state === STATES.REGEX) {
+      result = parseRegex(input, i);
+
+      if(result.error !== undefined) {
+        error = result.error;
```

Comment:
nit, I think early return on all errors would be slightly cleaner since the rest isn't going to matter anyways.